

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Quickstart &mdash; Bvalcalc 0.2.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=463a749d" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=37f418d5"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Tailoring Parameters" href="params.html" />
    <link rel="prev" title="Calculate Genome B-map" href="../modes/genome.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            Bvalcalc
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../introduction/cli_overview.html">CLI Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/generate_params.html">Generating parameters</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modes/site.html">Calculate Single Site B</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modes/gene.html">Calculate Gene B Recovery</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modes/region.html">Calculate Region B-map</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modes/genome.html">Calculate Genome B-map</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Guides</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#specifying-popgen-parameters">Specifying popgen parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#calculating-b-recovery-from-a-single-element">Calculating B recovery from a single element</a></li>
<li class="toctree-l2"><a class="reference internal" href="#calculating-b-for-a-region-of-the-genome">Calculating B for a region of the genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="#calculating-a-complete-b-map">Calculating a complete B-map</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="params.html">Tailoring Parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="selfing.html">Selfing</a></li>
<li class="toctree-l1"><a class="reference internal" href="demography.html">Demography</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Bvalcalc</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Quickstart</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/guides/quickstart.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="quickstart">
<h1>Quickstart<a class="headerlink" href="#quickstart" title="Link to this heading"></a></h1>
<section id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Link to this heading"></a></h2>
<p>You can install Bvalcalc via pip:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>bvalcalc
</pre></div>
</div>
</section>
<section id="specifying-popgen-parameters">
<h2>Specifying popgen parameters<a class="headerlink" href="#specifying-popgen-parameters" title="Link to this heading"></a></h2>
<p>First we need to copy population genetic parameters from one of the templates, e.g. Drosophila.
In your own analysis you’ll need to check the literature and use informed values for your population.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Bvalcalc<span class="w"> </span>--generate_params<span class="w"> </span>drosophila
</pre></div>
</div>
</section>
<section id="calculating-b-recovery-from-a-single-element">
<h2>Calculating B recovery from a single element<a class="headerlink" href="#calculating-b-recovery-from-a-single-element" title="Link to this heading"></a></h2>
<p>Now, let’s calculate the recovery of B as a function of distance from a single conserved element [10kb by default] and save it to gene_B.png</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Bvalcalc<span class="w"> </span><span class="se">\</span>
--gene<span class="w"> </span><span class="se">\</span>
--pop_params<span class="w"> </span>./DrosophilaParams.py<span class="w"> </span><span class="se">\</span>
--plot_output<span class="w"> </span>gene_B.png
</pre></div>
</div>
</section>
<section id="calculating-b-for-a-region-of-the-genome">
<h2>Calculating B for a region of the genome<a class="headerlink" href="#calculating-b-for-a-region-of-the-genome" title="Link to this heading"></a></h2>
<p>Next, we can use a BED file that specifies which genomic ranges are conserved to calculate B for a region in the genome.
Let’s calculate B for a 1Mb region in the middle of chromosome 2R [9500000-10500000] and save it to 1Mb_B.png.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Bvalcalc<span class="w"> </span><span class="se">\</span>
--region<span class="w"> </span>2R:9500000-10500000<span class="w"> </span><span class="se">\</span>
--pop_params<span class="w"> </span>./DrosophilaParams.py<span class="w"> </span><span class="se">\</span>
--bedgff_path<span class="w"> </span>examples/dmel6_2R_genes.csv<span class="w"> </span><span class="se">\</span>
--plot_output<span class="w"> </span>1Mb_B.png
</pre></div>
</div>
<p>Have a look at the plots, the blue sections of the graph indicate neutral regions and black indicates conserved elements.
That’s all that’s necessary for many analyses, especially if you’re only interested in B values for a specific region of the genome, or are testing against simulated results.</p>
</section>
<section id="calculating-a-complete-b-map">
<h2>Calculating a complete B-map<a class="headerlink" href="#calculating-a-complete-b-map" title="Link to this heading"></a></h2>
<p>If you wanted to generate a complete B-map for all sites across all chromosomes you can use the following command, though note its a lot more data to crunch!
The following is the basic command, recommended options include <code class="docutils literal notranslate"><span class="pre">--chr_sizes</span></code> with a file specifying each chromosome’s length, <code class="docutils literal notranslate"><span class="pre">--pop_change</span></code> for demography, <code class="docutils literal notranslate"><span class="pre">--rec_map</span></code> and <code class="docutils literal notranslate"><span class="pre">--gc_map</span></code> for crossover and gene conversion maps.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Bvalcalc<span class="w"> </span><span class="se">\</span>
--genome<span class="w"> </span><span class="se">\</span>
--pop_params<span class="w"> </span>./DrosophilaParams.py<span class="w"> </span><span class="se">\</span>
--bedgff_path<span class="w"> </span>examples/dmel6_2R_genes.csv
</pre></div>
</div>
<p>A caveat to the <code class="docutils literal notranslate"><span class="pre">--region</span></code> and <code class="docutils literal notranslate"><span class="pre">--genome</span></code> modes is that by default they combine and simplify distant elements in discrete chunks which can slightly change the distance of distant conserved elements when
calculating B. The default chunk size is 20kb and the window within which calculations are perfectly precise is three chunks in each direction (140kb total). This allows for vastly improved performance
and typically will not result in directional biases of B estimates. However, to achieve more exact results you can specify the size of the chunks with <code class="docutils literal notranslate"><span class="pre">--chunk_size</span></code>, and the size of the window to
perform perfectly precise calculations with <code class="docutils literal notranslate"><span class="pre">--precise_chunks</span></code>, though this will come at the cost of perfomance so consider using HPC resources or limiting to a specific region with <code class="docutils literal notranslate"><span class="pre">--region</span></code>.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../modes/genome.html" class="btn btn-neutral float-left" title="Calculate Genome B-map" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="params.html" class="btn btn-neutral float-right" title="Tailoring Parameters" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Jacob Marsh, Parul Johri.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>