


<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Quickstart Tutorial &mdash; Bvalcalc 1.3.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=7b5bf6d2" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=1f29e9d3"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/js/onthispage-scrollspy.js?v=e5c7a674"></script>
      <script src="../_static/js/interactive-code.js?v=6de093d8"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Tailoring Parameters" href="params.html" />
    <link rel="prev" title="B-map Utilities for VCF" href="../modules/bmap_utils.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            Bvalcalc
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../introduction/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/bedgff_input.html">BED/GFF Input</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction/generate_params.html">Generate Parameters</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/site.html">Calculate Single Site B</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/gene.html">Calculate Gene B Recovery</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/region.html">Calculate Region B-map</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genome.html">Calculate Genome B-map</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/bmap_utils.html">B-map Utilities for VCF</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Guides</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quickstart Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="params.html">Tailoring Parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="multiple_dfes.html">Multiple DFEs</a></li>
<li class="toctree-l1"><a class="reference internal" href="demography.html">Demographic Inference with B-map</a></li>
<li class="toctree-l1"><a class="reference internal" href="sweepfinder2.html">SweepFinder2 with B-map</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api/calculateB.html">Calculate B</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">About BGS and B-values</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../prebuilt_maps/learn.html">Work in Progress…</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Query species B-maps</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../prebuilt_maps/query.html">Work in Progress…</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Bvalcalc</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Quickstart Tutorial</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="quickstart-tutorial">
<h1>Quickstart Tutorial<a class="headerlink" href="#quickstart-tutorial" title="Link to this heading"></a></h1>
<p>This tutorial (5-10 minutes) walks through the core functionality of calculating B with <strong>Bvalcalc</strong>!</p>
<section id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Link to this heading"></a></h2>
<p><strong>Bvalcalc</strong> is a python program called from the command line (terminal, shell, zsh, bash etc.)</p>
<p>Install <strong>Bvalcalc</strong> in a new conda environment via conda-forge:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>bvalcalc<span class="w"> </span>-c<span class="w"> </span>conda-forge<span class="w"> </span>bvalcalc
conda<span class="w"> </span>activate<span class="w"> </span>bvalcalc
</pre></div>
</div>
</section>
<section id="specifying-popgen-parameters">
<h2>Specifying popgen parameters<a class="headerlink" href="#specifying-popgen-parameters" title="Link to this heading"></a></h2>
<p>To calculate B, we need evolutionary information about the species of the population in a parameters file.</p>
<p>Copy population genetic parameters from one of the templates, e.g. for coding (CDS) regions in <cite>Drosophila melanogaster</cite> (DroMel).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Bvalcalc<span class="w"> </span>--generate_params<span class="w"> </span>dromel_cds

<span class="c1"># Wrote parameters to: ./DroMel_Cds_Params.py</span>
</pre></div>
</div>
<p>Open <code class="docutils literal notranslate"><span class="pre">./DroMel_Cds_Params.py</span></code> in your text editor of choice. This file contains popgen parameters that allow us to accurately calculate B.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Templates are available for several species and annotation types, see the list at <a class="reference internal" href="../introduction/generate_params.html"><span class="doc">Generate Parameters</span></a>.
When calculating B for a new species or annotation type you’ll need to check the literature and use inferred/informed values for your population, see <a class="reference internal" href="params.html"><span class="doc">Tailoring Parameters</span></a>.</p>
</div>
</section>
<section id="calculating-a-b-value">
<h2>Calculating a B value<a class="headerlink" href="#calculating-a-b-value" title="Link to this heading"></a></h2>
<p>Let’s start by calculating B at a site 1kb from the edge of a single conserved element (CDS) 5kb in length.</p>
<p>A conserved element is an annotated part of the genome that experiences direct purifying selection. It will introduce BGS (reduction in B) at sites within and outside the element.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Bvalcalc<span class="w"> </span><span class="se">\</span>
--site<span class="w"> </span><span class="se">\</span>
--params<span class="w"> </span>./DroMel_Cds_Params.py<span class="w"> </span><span class="se">\</span>
--element_size<span class="w"> </span><span class="m">5000</span><span class="w"> </span><span class="se">\</span>
--distance<span class="w"> </span><span class="m">1000</span>

<span class="c1"># B for site 1000bp away from 5000bp region: 0.9868859634102181</span>
</pre></div>
</div>
<p>The B-value for this site should be printed in the console as ~0.987, which indicates diversity at your focal site is reduced ~1.3% by BGS from the 5kb conserved element.</p>
</section>
<section id="b-recovery-from-one-element">
<h2>B recovery from one element<a class="headerlink" href="#b-recovery-from-one-element" title="Link to this heading"></a></h2>
<p>Now, let’s calculate the recovery of B as a function of distance from the 5kb element so we can plot it (<code class="docutils literal notranslate"><span class="pre">gene_B.png</span></code>).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Bvalcalc<span class="w"> </span><span class="se">\</span>
--gene<span class="w"> </span><span class="se">\</span>
--element_size<span class="w"> </span><span class="m">5000</span><span class="w"> </span><span class="se">\</span>
--params<span class="w"> </span>./DroMel_Cds_Params.py<span class="w"> </span><span class="se">\</span>
--plot<span class="w"> </span>gene_B.png

<span class="c1"># B for adjacent site: 0.9554782496927732</span>
<span class="c1"># Mean B for flanking region: 0.9982138514437486</span>
</pre></div>
</div>
<p>Have a look at the plot and the results printed to the console, you’ll notice B decays with distance from the selected element. It’s a modest reduction, but remember, across a genome ALL selected elements will contribute to B at any given site.</p>
<img alt="B gene tutorial example" class="with-shadow align-center" src="../_images/gene_B_tutorial.png" />
</section>
<section id="b-in-a-genomic-region">
<h2>B in a genomic region<a class="headerlink" href="#b-in-a-genomic-region" title="Link to this heading"></a></h2>
<p>Alright, now let’s look at part of a chromosome.
We can use a BED file (or GFF/CSV) that specifies which genomic ranges are conserved to calculate B for a region in the genome. Here we’ll also provide an optional recombination (crossover) map to account for local variation in crossover rates.</p>
<p>First we’ll need to download the sample data, which includes a BED file with CDS annotations and a recombination map for <cite>D. melanogaster</cite>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Bvalcalc<span class="w"> </span>--download_sample_data
</pre></div>
</div>
<p>Let’s calculate B for a 1 Mb region in a relatively low recombination region in chromosome 2R and save it to <code class="docutils literal notranslate"><span class="pre">1Mb_B.png</span></code> (to save the B-map output to a file, use <code class="docutils literal notranslate"><span class="pre">--out</span></code>).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Bvalcalc<span class="w"> </span><span class="se">\</span>
--region<span class="w"> </span>chr2R:8000000-9000000<span class="w"> </span><span class="se">\</span>
--params<span class="w"> </span>./DroMel_Cds_Params.py<span class="w"> </span><span class="se">\</span>
--bedgff<span class="w"> </span>./cds_noX.bed<span class="w"> </span><span class="se">\</span>
--rec_map<span class="w"> </span>./dmel_comeron_recmap.csv<span class="w"> </span><span class="se">\</span>
--plot<span class="w"> </span>1Mb_B.png

<span class="c1"># Mean B of neutral sites across specified region: 0.8925140916007113</span>
<span class="c1"># Plot saved to 1Mb_B.png</span>
<span class="c1"># = B value calculated in 6.67 seconds. = = =</span>
</pre></div>
</div>
<p>Have a look at the plot: the blue sections of the graph indicate neutral regions and black indicates conserved elements (CDS). The white-red gradient reflects the local crossover rate (red = high).</p>
<img alt="B region tutorial example" class="with-shadow align-center" src="../_images/1Mb_B.png" />
<p>That’s all that’s necessary for many analyses, especially if you’re only interested in B values for a specific region of the genome, or are testing against simulated results.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The recombination map directly scales the crossover rate <code class="docutils literal notranslate"><span class="pre">r</span></code>.
You could also add a gene conversion map with <code class="docutils literal notranslate"><span class="pre">--gc_map</span></code> but as the gc rate is relatively constant in <cite>D. melanogaster</cite> (Comeron et al. 2012), you can calculate a reasonably accurate map without it.</p>
</div>
</section>
<section id="calculating-a-b-map">
<h2>Calculating a B-map<a class="headerlink" href="#calculating-a-b-map" title="Link to this heading"></a></h2>
<p>If you wanted to generate a complete B-map for all sites across all chromosomes you would use the following command, though note it’s a lot more data to crunch and maps are already available for Drosophila so no need to run it!</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Bvalcalc<span class="w"> </span><span class="se">\</span>
--genome<span class="w"> </span><span class="se">\</span>
--params<span class="w"> </span>./DroMel_Cds_Params.py<span class="w"> </span><span class="se">\</span>
--bedgff<span class="w"> </span>./cds_noX.bed<span class="w"> </span><span class="se">\</span>
--rec_map<span class="w"> </span>./dmel_comeron_recmap.csv<span class="w"> </span><span class="se">\</span>
--out<span class="w"> </span>Dmel_Bmap.csv<span class="w"> </span><span class="se">\</span>
--out_binsize<span class="w"> </span><span class="m">1000</span>
</pre></div>
</div>
<p>If you had run that command, you’d get a B-map for sites due to BGS from CDS regions. It’s also possible to include the contribution of BGS from non-coding regions using <strong>Bvalcalc</strong> (see <a class="reference internal" href="multiple_dfes.html"><span class="doc">Multiple DFEs</span></a>).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>B-maps are useful to identify highly conserved regions of the genome, as a null-model for inference, e.g. <a class="reference internal" href="sweepfinder2.html"><span class="doc">SweepFinder2 with B-map</span></a>, or to select the most neutrally-evolving sites for e.g. demographic inference, see <a class="reference internal" href="demography.html"><span class="doc">Demographic Inference with B-map</span></a>.</p>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../modules/bmap_utils.html" class="btn btn-neutral float-left" title="B-map Utilities for VCF" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="params.html" class="btn btn-neutral float-right" title="Tailoring Parameters" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Jacob Marsh, Parul Johri.</p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(false);
      });
  </script>
   
  <!-- To bottom button -->
  <div class="scroll-bottom-wrapper">
    <button class="scroll-to-bottom-btn" aria-label="Scroll to bottom">
      <span class="scroll-to-bottom-btn-text">↓ ↓ ↓</span>
    </button>
  </div>
  <!-- Right elements -->
  <div class="right-elements-container">
    <div class="on-this-page">
        <ul>
<li><a class="reference internal" href="#">Quickstart Tutorial</a><ul>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#specifying-popgen-parameters">Specifying popgen parameters</a></li>
<li><a class="reference internal" href="#calculating-a-b-value">Calculating a B value</a></li>
<li><a class="reference internal" href="#b-recovery-from-one-element">B recovery from one element</a></li>
<li><a class="reference internal" href="#b-in-a-genomic-region">B in a genomic region</a></li>
<li><a class="reference internal" href="#calculating-a-b-map">Calculating a B-map</a></li>
</ul>
</li>
</ul>

    </div>
    <img src="../_static/Bvalcalc_bee_trail.svg" aria-hidden="true" class="bee-trail-img" />
    <div class="bottom-right-elements">
      <div class="citation-text">
        <strong>Citation:</strong><br/>Marsh, JI, Johri, PJ (2025).
        <span class="citation-body">
          The B-value calculator: map expected diversity
          under background selection. <i>bioRxiv</i>. 
        </span>
        <a href="https://biorxiv" target="_blank" rel="noopener noreferrer">10.1101/2063.10.10.561727</a>
      </div>      
      <div class="made-by-sphinx-text">
        <a href="https://www.sphinx-doc.org/" target="_blank" rel="noopener noreferrer">Sphinx</a> | 
        <a href="https://readthedocs.org/" target="_blank" rel="noopener noreferrer">Read the Docs</a> <br/>
      </div>
      <div class="links-text">
        <a href="https://pypi.org/project/bvalcalc/" target="_blank" rel="noopener noreferrer">PyPI</a> | 
        <a href="https://github.com/JohriLab/Bvalcalc/issues" target="_blank" rel="noopener noreferrer">Report a bug</a> | 
        <a href="https://github.com/JohriLab/Bvalcalc" target="_blank" rel="noopener noreferrer">Source code</a> 
      </div>
    </div>  
  </div>


</body>
</html>